File: android-offline/src/com/apolos/shieldlite/SecurityState.java
package com.apolos.shieldlite;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

/** Process-wide, thread-safe security state shared between the service and the UI. */
public final class SecurityState {

    public static final int SEV_INFO = 0;
    public static final int SEV_WARNING = 1;
    public static final int SEV_CRITICAL = 2;

    public static final class Event {
        public final long id;
        public final int severity;
        public final String title;
        public final String detail;
        public final long timestamp;

        Event(long id, int severity, String title, String detail) {
            this.id = id;
            this.severity = severity;
            this.title = title;
            this.detail = detail;
            this.timestamp = System.currentTimeMillis();
        }
    }

    public interface Listener {
        void onChanged();
    }

    private static final AtomicLong ID_GEN = new AtomicLong(0);
    private static final Object LOCK = new Object();
    private static final List<Event> EVENTS = new ArrayList<>();
    private static final List<Listener> LISTENERS = new ArrayList<>();
    private static final int MAX_EVENTS = 200;

    private static boolean monitoringActive = false;
    private static boolean vpnActive = false;
    private static String vpnMode = "off";
    private static boolean cameraInUse = false;
    private static boolean micInUse = false;
    private static long blockedDnsCount = 0;
    private static Event alarm = null;

    private SecurityState() { }

    public static void addListener(Listener l) {
        synchronized (LOCK) { LISTENERS.add(l); }
    }

    public static void removeListener(Listener l) {
        synchronized (LOCK) { LISTENERS.remove(l); }
    }

    private static void notifyListeners() {
        List<Listener> copy;
        synchronized (LOCK) { copy = new ArrayList<>(LISTENERS); }
        for (Listener l : copy) l.onChanged();
    }

    public static Event addEvent(int severity, String title, String detail) {
        Event e = new Event(ID_GEN.incrementAndGet(), severity, title, detail);
        synchronized (LOCK) {
            EVENTS.add(0, e);
            while (EVENTS.size() > MAX_EVENTS) EVENTS.remove(EVENTS.size() - 1);
            if (severity == SEV_CRITICAL) alarm = e;
        }
        notifyListeners();
        return e;
    }

    public static void clearAlarm() {
        synchronized (LOCK) { alarm = null; }
        notifyListeners();
    }

    public static Event getAlarm() {
        synchronized (LOCK) { return alarm; }
    }

    public static List<Event> getEvents() {
        synchronized (LOCK) { return new ArrayList<>(EVENTS); }
    }

    public static void setMonitoringActive(boolean v) {
        synchronized (LOCK) { monitoringActive = v; }
        notifyListeners();
    }

    public static boolean isMonitoringActive() {
        synchronized (LOCK) { return monitoringActive; }
    }

    public static void setVpn(boolean active, String mode) {
        synchronized (LOCK) { vpnActive = active; vpnMode = mode; }
        notifyListeners();
    }

    public static boolean isVpnActive() {
        synchronized (LOCK) { return vpnActive; }
    }

    public static String getVpnMode() {
        synchronized (LOCK) { return vpnMode; }
    }

    public static void setCameraInUse(boolean v) {
        synchronized (LOCK) { cameraInUse = v; }
        notifyListeners();
    }

    public static boolean isCameraInUse() {
        synchronized (LOCK) { return cameraInUse; }
    }

    public static void setMicInUse(boolean v) {
        synchronized (LOCK) { micInUse = v; }
        notifyListeners();
    }

    public static boolean isMicInUse() {
        synchronized (LOCK) { return micInUse; }
    }

    public static void incrementBlockedDns() {
        synchronized (LOCK) { blockedDnsCount++; }
        notifyListeners();
    }

    public static long getBlockedDnsCount() {
        synchronized (LOCK) { return blockedDnsCount; }
    }

    /** INFO / WARNING / CRITICAL summary the dashboard colours itself by. */
    public static int threatLevel() {
        synchronized (LOCK) {
            if (cameraInUse || micInUse) return SEV_CRITICAL;
            if (!vpnActive) return SEV_WARNING;
            return SEV_INFO;
        }
    }
}

File: android/app/src/main/java/com/apolos/shield/core/SecurityState.kt
package com.apolos.shield.core

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import java.util.concurrent.atomic.AtomicLong

/** Severity of a security event, drives the dashboard colour + alarm. */
enum class Severity { INFO, WARNING, CRITICAL }

/** Category of the sensor / subsystem that produced an event. */
enum class EventKind { CAMERA, MICROPHONE, SCREEN, APP, TRAFFIC, VPN, SYSTEM }

data class SecurityEvent(
    val id: Long,
    val kind: EventKind,
    val severity: Severity,
    val title: String,
    val detail: String,
    val timestamp: Long = System.currentTimeMillis(),
)

/** Snapshot of an installed app that the behaviour monitor flagged. */
data class AppRisk(
    val packageName: String,
    val label: String,
    val riskScore: Int,          // 0..100
    val reasons: List<String>,
    val sideloaded: Boolean,
)

/** Live per-subsystem status shown on the dashboard. */
data class ShieldStatus(
    val vpnActive: Boolean = false,
    val vpnMode: String = "off",           // off | killswitch | dns | wireguard
    val monitoringActive: Boolean = false,
    val cameraInUse: Boolean = false,
    val micInUse: Boolean = false,
    val screenCaptured: Boolean = false,
    val blockedDnsCount: Long = 0,
    val flaggedApps: Int = 0,
    val threat: Severity = Severity.INFO,
)

/**
 * Process-wide, thread-safe source of truth. Every monitor writes here and the
 * Compose UI + Notifier observe it. Kept as a singleton object so the foreground
 * service and the Activity share one instance.
 */
object SecurityState {

    private val idGen = AtomicLong(0)

    private val _status = MutableStateFlow(ShieldStatus())
    val status: StateFlow<ShieldStatus> = _status.asStateFlow()

    private val _events = MutableStateFlow<List<SecurityEvent>>(emptyList())
    val events: StateFlow<List<SecurityEvent>> = _events.asStateFlow()

    private val _apps = MutableStateFlow<List<AppRisk>>(emptyList())
    val apps: StateFlow<List<AppRisk>> = _apps.asStateFlow()

    /** Raised to true whenever a CRITICAL event fires; UI shows the alarm banner. */
    private val _alarm = MutableStateFlow<SecurityEvent?>(null)
    val alarm: StateFlow<SecurityEvent?> = _alarm.asStateFlow()

    fun updateStatus(transform: (ShieldStatus) -> ShieldStatus) {
        _status.update(transform)
        recomputeThreat()
    }

    fun addEvent(kind: EventKind, severity: Severity, title: String, detail: String): SecurityEvent {
        val event = SecurityEvent(idGen.incrementAndGet(), kind, severity, title, detail)
        _events.update { (listOf(event) + it).take(MAX_EVENTS) }
        if (severity == Severity.CRITICAL) {
            _alarm.value = event
        }
        recomputeThreat()
        return event
    }

    fun setApps(list: List<AppRisk>) {
        _apps.value = list
        _status.update { it.copy(flaggedApps = list.count { a -> a.riskScore >= 60 }) }
        recomputeThreat()
    }

    fun clearAlarm() { _alarm.value = null }

    private fun recomputeThreat() {
        val recentCritical = _events.value.take(5).any { it.severity == Severity.CRITICAL }
        _status.update { s ->
            val activeCritical = s.cameraInUse || s.micInUse || s.screenCaptured
            val threat = when {
                activeCritical || recentCritical -> Severity.CRITICAL
                s.flaggedApps > 0 || !s.vpnActive -> Severity.WARNING
                else -> Severity.INFO
            }
            if (threat == s.threat) s else s.copy(threat = threat)
        }
    }

    private const val MAX_EVENTS = 200
}
